### ISSUE TITLE
Measuring the quality of different SHAP imputers

### ISSUE THREAD

**Issue by adrian-prior** [PriorLabs/tabpfn-extensions#356]
Hello everyone,

After looking into which approximators would work best for our SHAP-value computations in PriorLabs/tabpfn-extensions#342, I'm now wondering whether there's a way to evaluate the quality of the imputers at our disposal and make an informed choice for our default based on this.

My basic idea for this was the following (bear with me, I am really not a SHAP expert, so please do criticize). The value function we want to approximate with our imputer is $v(S)=E_q[f(X)|X_s=x_S]$, where $q=p(x_\bar{S})$ (marginal) or $q=p(x_\bar{S}|x_S)$ (conditional). If I understand it correctly, the choice of $q$ depends on which formulation of the underlying game we chose to compute the SHAP values, but for the general case, where we want to explain the model behavior vs. the data, the marginal formulation is preferable.

$v(S)$ can be pretty well estimated by just marginalizing over the training set, i.e., just taking the mean over $f(x_S, x_{i, \bar{S}})$ for all $x_i$ in the training set, so that could serve as a proxy for what a "good" imputation looks like. We just normally don't marginalize over the training set, because this requires $n_\text{train}$ model predictions.

Equipped with this proxy ground truth, we can now measure the quality of different imputers on different datasets.

Below you can find some details on the imputers and datasets I tested (mostly generated with CC):

<details><summary>Imputers evaluated</summary>

Versions: `shapiq==1.6.0`, `tabpfn-extensions` @ `d7e7110` (origin/main, has `get_tabpfn_inf_explainer`), `tabpfn==8.1.0`.

| imputer | how it fills absent features | passes | family | implementation |
| -- | -- | -- | -- | -- |
| **mean / baseline** | per-feature train mean (one fixed value) | 1 | marginal | shapiq `BaselineImputer`, via tabpfn-extensions `get_tabpfn_imputation_explainer(imputer="baseline")` — **unmodified** |
| **inf** | set to `+inf`; TabPFN's native missing-handling absorbs them | 1 | conditional | tabpfn-extensions `_InfImputer` (nested in `get_tabpfn_inf_explainer`) — **unmodified** (used via `explainer._imputer`) |
| **nan_fill** | set to `NaN`; handled by TabPFN preprocessing (sanity-check floor) | 1 | marginal (TabPFN preprocessing imputes NaNs before they reach the model) | **ours** — no shipped NaN imputer exists, so a 1-line `MarginalImputer` subclass mirroring `_InfImputer` with `np.nan` instead of `+inf` |
| **marginal (sample_size sweep)** | average over `s` random train rows substituted into absent features; `s ∈ {1..500}` | s | marginal | **ours** — `BatchedMarginalImputer`, a `shapiq.MarginalImputer` subclass that batches all samples into a few predict calls instead of shapiq's per-sample Python loop. Reuses the parent's sampling + empty-prediction ⇒ **numerically identical (< />1e-4), < />49× faster** |
| **conditional** | rows drawn from x's neighborhood (tree-embedding + Hamming distance), `sample_size=10` | ~10 | conditional | shapiq `GenerativeConditionalImputer`, via `get_tabpfn_imputation_explainer(imputer="conditional")` — **unmodified** (default params) |

Notes:

* **Ground truth** = batched marginalization over the *full* train set — equivalent to shapiq's `MarginalImputer` at `sample_size = n_train` (cross-checked to `max|diff| ≈ 1e-4`), just batched for speed.
* All imputers *and* the ground truth share the exact same model output via shapiq's `get_predict_function_and_model_type` (prediction for regression, class-1 **logit** for classification).

</details>

<details><summary>Datasets (7)</summary>

| dataset | task | n_features | notes |
| -- | -- | -- | -- |
| synthetic | regression | 20 | linear toy: `y = x0 + 0.5·x1 + noise`, 18 noise features, independent |
| friedman1 | regression | 20 | `make_friedman1` — non-linear (sin + squared + interaction) |
| communities_and_crime | regression | 122 | OpenML 42730 |
| online_news_popularity | regression | 58 | OpenML 4545 |
| spambase | classification (binary) | 57 | OpenML 44 |
| magic_telescope | classification (binary) | 10 | OpenML 1120 |
| synthetic_classification | classification (binary) | 20 | `make_classification` — 5 redundant (correlated) feats, 2 clusters/class, 2% label noise |

100 test instances/dataset; coalitions stratified by fraction-of-d. Public `tabpfn` (cache-capable stack the imputers target).

</details>

Note that I ran the following experiments before I really understood the distinction between conditional and marginal case, **so it's clear that some of the imputers perform worse just because they measure something different than the proxy ground truth set out. **

Here are some initial results though:

**Quality–time trade-off (per dataset; marginal curve labeled with # train samples marginalized over)**

<img src="https://github.com/user-attachments/assets/0b38abfe-86cb-4f49-a1f4-43183905b8f3 " alt="Image" width="4104" data-linear-height="3745" />

**Error vs coalition size (per dataset)**

<img src="https://github.com/user-attachments/assets/8f025aab-f9e3-4940-aa9d-00830a613cc8 " alt="Image" width="4112" data-linear-height="3752" />

**Normalized error summary (error / per-dataset ground-truth std, lower = better)**

<img src="https://github.com/user-attachments/assets/eb827602-1403-426f-8f41-d26d2a0a92f9 " alt="Image" width="4532" data-linear-height="1231" />

I was thinking that before I continue at all with this line of work, I wanted to get the opinion of some people more familiar with this topic.  <br>@mmschlk @Advueu963 Maybe you have some opinions regarding this approach.

**Here are just a few questions that I was asking myself (feel free to ignore and just give your general opinion):**

* Does it make any sense?
* Is my call on marginal vs. conditional correct if we care about explaining model performance?
* Are there any alternative approaches for measuring the quality of imputers for computing SHAP values? E.g., in this approach, it still isn't clear at all how much the "quality" of the final SHAP values is affected.
* At least with the marginal imputer, you can trade off cost against quality. In general, a similar trade-off/knob exists with the number of coalitions to sample. Do you have any hunch which knob is more important/how to trade these off against each other?

Many thanks in advance! Any feedback is greatly appreciated!

**Comment by mmschlk** [comment]
Hey @adrian-prior, thanks for reaching out! Generally, yes it makes sense, but having a measurable quantity is really hard.

**TLDR**: Deciding what "best" means is really hard (still open research) and context-dependent. Yet, imputation is not the thing we want in explanation but a means to an end. Model's that _know_ what missingess means should (in my opinion) always be explained with this mechanism and not run through some convoluted imputation scheme (if your explanation goal does not need it).

First and foremost, the decision of what imputation is the right one is not one that you can simply say as there really is a difference in what your final explanations will tell you **[[1](https://arxiv.org/abs/2006.16234),[2](https:/arxiv.org/abs/1903.10464), and many more]** . Explaining by breaking the dependencies in the data (imputation with marginal) is more explaining the model. Explaining by being really careful to not do this and keep the data distribution intact (imputation with conditional) explains the model and the data generating process. Both of these explanations can vary drastically based on mainly two mechanisms at play: interactions in the model or correlated features. If you explain true to the model (marginal) you will spot the weird interactions that the model has picked up more clearly. If you explain true to the data (conditional) you can get attribution scores that tell you what the explanatory power is in the features and the dataset but this will often hide the quirky things actually learned by the model. So yes, going with the "marginal" view is preferred if you want to learn stuff about your model!

However, **imputation is actually just a means to an end** and not really the thing we want to do for getting good explanations. What you want to do for explaining your model is to **remove information from the model**, see how it performs without it, and then do this many times to aggregate your findings into summary scores (Shapley Values, interactions, feature importance, effect curves, and many more XAI measures fall into this). So it's all about _removing_ information. If your model already knows what "missigness" is then there actually already exist a good solution to your explanation target: just tell the model you do not know what this feature value is supposed to be. For example in trees, many variants have a "NA prediction value on decision nodes for the case if a feature is missing" or if does not then we can just say we do not know what a particular feature is I am encountering at this decision node, so I will walk both ways down the tree and see where this leads me and average the results into the prediction of the model. This is the de-facto correct way of explaining trees. TabPFN does this mechanism as well, so I would also explain with this in mind. However, I there is one more piece: You can pass your internal NA value which will signal to the model idk what this is supposed to be and it will do whatever you guys have put into it **or** you could just do not provide the feature column entirely [**[2]**](https://link.springer.com/chapter/10.1007/978-3-031-63797-1_23). This is what we named remove-and-recontextualize and is implemented in the current TabPFN imputer. If the model has no knowledge about your feature it cannot do anything with it. This is different from knowing there is a feature "longitude" but I do not know the value of it. Again, two different interpretations for explanations. Two different interpretations that have both their merit for one question or another. I would argue the "remove the feature column entirely" is the correct way of doing this, which might be getting hard to cache though.

If you really wanna go down the marginal imputation route (while I am about 80% confident that TabPFN should not be explained like this) then you could check out CTE (compress-then-explain) [**[3]**](https://openreview.net/forum?id=LiUfN9h0Lx&noteId=ZLTc93e9SF) by my friend Hubert. He also asked what data should we use to make explanations and showed that the "best" solution for the marginal case is compressing the data distribution of the background samples and drawing just a couple of samples from this instead of sampling too much. This is also on the way for shapiq of course.

**For your questions:**
- If you want to do some investigation then I would not look at the error on the value function $\nu(S)$ but on the explanations $\phi$ (the shap scores) kind-of like Hubert did in [**[3]**](https://openreview.net/forum?id=LiUfN9h0Lx&noteId=ZLTc93e9SF) but this again is actually very hard to do a cross-imputer comparison. 
- **Knob**: Again, since there is no clear "better" definition when it comes to the strategy its a bit hard to say quality vs cost here, but generally I would love to have more coalitions evaluated than high-fidelity estimates about the model average. Baseline imputers / native-missingnes imputers do not really have a knob.
- **Alternatives**: we always make nice synthetic functions we know how the explanations should look like and see how the explanations capture this. but this only works with glass-box models, trees, linear, GLMs. A good proxy for black-box models is measuring their faithfulness: Compute explanations -> derive some expected behavior of the model from the explanations -> see how good your expectation was. This is often measured in terms of insertion deletion curves [**[4, examples]**](https://openreview.net/forum?id=on22Rx5A4F)

**Comment by mmschlk** [comment]
**shapiq-v2:** A quick thing though: I think this whole discussion is super shitty in current XAI libraries. It's one of the core decisions influencing your interpretation, but it's nowhere well enforced nor informed (usually just some option that suggest all are equally important). shapiq is no exception. However, we are currently rewriting shapiq from the ground up with so many learnings from all of the last years in research to make it actually really easy _but also necessary_ to choose "what do I want to explain" first and then run it through the explanation pipeline. Better PFN support is of course one of our things heavily influencing our core API design and choices.

It's quite hard to condense all the different things into a comment. We would love to have a chat with you guys. @Advueu963 and I have quite some substantial changes in the works where we would really appreciate also your guys input!

**Comment by adrian-prior** [comment]
@mmschlk,

Thanks a lot for this detailed response. There is so much nuance in this topic, and it really feels like I am just starting to scratch the surface, so this is super valuable for us. I totally agree that a GitHub issue isn't the right format for this discussion (without it taking months), and we would love to hear more about the **shapiq-v2** rewrite you have in the works. I will email @Advueu963 and you to set up a meeting!

PS: I might post some follow-up questions here if/when I have a chance to dive deeper into the literature!

**Comment by mmschlk** [comment]
Yes! The field in my opinion has in itself only recently really understood what it wants to do with these game theoretic values and perspectives on the models and everything is always a bit interlinked. Feel free to raise any follow-up questions, of course. :)
